ledger-core/v2/network/consensus/adapters/tests/pulse_test.go

```diff
@@ -56,19 +74,26 @@ func (p *Pulsar) Pulse(ctx context.Context, attempts int) {
 
 	data := pulse.NewPulsarData(p.pulseNumber, p.pulseDelta, prevDelta, randBits256())
 	p.pulseNumber += pulse.Number(p.pulseDelta)
-
-	pu := adapters.NewPulse(data)
-	ph, _ := host.NewHost("127.0.0.1:1")
-	th, _ := net.ResolveTCPAddr("tcp", "127.0.0.1:2")
-	pp := pulsenetwork.NewPulsePacketWithTrace(ctx, &pu, ph, th, 0)
-
-	bs, _ := packet.SerializePacket(pp)
-	rp, _, _ := packet.DeserializePacketRaw(bytes.NewReader(bs))
+	pp := serialization.BuildPulsarPacket(ctx, data)
 
 	go func() {
 		for i := 0; i < attempts; i++ {
-			handler := p.pulseHandlers[rand.Intn(len(p.pulseHandlers))]
-			go handler.HandlePulse(ctx, pu, rp)
+			nodeID := rand.Intn(len(p.addresses))
+			address := p.addresses[nodeID]
+			transport := p.transports[nodeID]
+			go func() {
```

Comment:
This is a potential resource leak. I suggest to consider limiting the total number of goroutines at least somehow, e.g. use an atomic variable that limits the total amount of goroutines to, say, 1000. A metric would be nice too, although this might be not a priority for now.

Reply:
it's a test dummy pulsar, It shoots `attempts` goroutines per tick in tests. There are no infinite loops/infinite selects in goroutine so all of them will stop eventually. I don't see resource leak

Reply:
OK, I missed that it's in tests/ directory, sri.